src/graph.ts

```diff
@@ -0,0 +1,104 @@
+import * as path from 'path'
+import * as jsonParser from '@ts-common/json-parser'
+import * as fs from 'fs';
+import * as json from '@ts-common/json'
+import * as stringMap from '@ts-common/string-map'
+import * as it from '@ts-common/iterator'
+
+
+type Ref = {
+  /**
+   * URL of JSON document.
+   */
+  readonly url: string
+  /**
+   * JSON pointer.
+   */
+  readonly pointer: string
+}
+
+const getRefs = (j: json.Json): it.IterableEx<string> => {
+  if (json.isObject(j)) {
+    return stringMap
+      .entries(j)
+      .flatMap(([k, v]) => (k === '$ref' && typeof v === 'string' ? it.concat([v]) : getRefs(v)))
+  } else if (it.isArray(j)) {
+    return it.flatMap(j, getRefs)
+  } else {
+    return it.empty()
+  }
+}
+
+
+const parseRef = (ref: string): Ref => {
+  const i = ref.indexOf('#')
+  return i < 0 ? { url: ref, pointer: '' } : { url: ref.substr(0, i), pointer: ref.substr(i + 1) }
+}
+
+const getReferencedFileNames = (fileName: string, doc: json.Json) => {
+  const dir = path.dirname(fileName)
+  return getRefs(doc)
+    .map(v => parseRef(v).url)
+    .filter(u => u !== '')
+    .map(u => path.resolve(path.join(dir, u)))
+}
+
+export class Graph {
+  readonly graph: Map<string, string[]>

```

Comment:
Instead of `string[]` use `Document`. Where `Document` is `{ readonly references: () => Iterable<string> }`

Reply:
Another option to define the `Document` is 
```ts
type Document = {
   readonly filePath: string
   readonly references: Iterable<Document> 
}
```

Reply:
Targets:
1. make sure there is no circular references. If there are circular references, break a cycle and report an error.
2. `const getAllReferencedDocuments = (document: Document) => Iterable<Document>`. Remove duplication `it.uniq`.